### ISSUE TITLE
Expecting higher performance from select count(*) 

### ISSUE THREAD

**Issue by abaror** [partiql/partiql-lang-kotlin#147]
It is human visible slow with only ~100K docs.  I feel like the count(*) should be cached when you create the doc , or even simple loop over only top of the docs should not be as slow.  Suspicious that deep parsing is happening.

**Comment by dlurton** [comment]
It is true that `COUNT(*)` is not very optimized (at the moment) in the reference implementation of PartiQL.  Usually, however, this type of optimization is usually best applied as part of a subsystem which knows a lot more about the data storage system than PartiQL does.

For example, one way to go about optimizing this would be with an AST rewrite that changes queries like `SELECT COUNT(*) FROM someTable` to call a function that is provided by the application hosting PartiQL.  This function can looks up the number of rows in the table in an optimized fashion.  If expressed using PartiQL syntax, the rewritten query might look like: `$TABLE_ROW_COUNT('someTable')`.

**Comment by abaror** [comment]
I don't completly agree. My two cents:

Yes, I can cache the stats in my app and do lazy evaluation. This is exactly what query optimizers can do for you.  Many query systems store some stats along with the dataset which allows them to perform optimization. Cardinality is the first one to start with :) .  

The need to know the storage system is only if you assume you need to return result that is reflecting the current file which can change , and then we need to store hash etc.  However, if we only assume that we only care about memory , we can keep the stats along with the objects that wrap them and create them together when loading.

**Comment by dlurton** [comment]
For the PartiQL reference implementation the loading of data is done by the `Binding<>` and/or `ExprValue` implementations, both of which can be provided by the host application.  PartiQL is very storage agnostic on purpose and doesn't make any assumptions about what data can be cached or not.  PartiQL doesn't do any data caching at all.

I'm not convinced that the PartiQL reference implementation should do caching of any kind since we don't want to make any assumptions about the underlying storage mechanisms.

If you are inclined to make such an optimization yourself, we do have some building blocks you can use:

- [lazy bindings example](https://github.com/partiql/partiql-lang-kotlin/blob/master/examples/src/kotlin/org/partiql/examples/SimpleExpressionEvaluation.kt)
- [CSV ExprValue example](https://github.com/partiql/partiql-lang-kotlin/blob/master/examples/src/kotlin/org/partiql/examples/CsvExprValueExample.kt).  
- [partial evaluation example](https://github.com/partiql/partiql-lang-kotlin/blob/master/examples/src/kotlin/org/partiql/examples/PartialEvaluationRewriter.kt) is an example of an AST rewrite.

**Comment by abaror** [comment]
Ok, Now I understand better where you put the line. So the reference implementation does not even manage the in-memory store of the ExprValue.  You do it in the REPL , but that is just an example and you don’t assume that people will use that for optimized processing.

**Comment by dlurton** [comment]
Yes, the REPL in this case is a host application and is primarily intended for experimentation.
